zk/stages/stage_batches.go

```diff
@@ -293,7 +354,86 @@ LOOP:
 						// when the stage is fired up for the first time
 						log.Warn(fmt.Sprintf("[%s] Skipping block %d, already processed", logPrefix, entry.L2BlockNumber))
 					}
-					continue
+
+					dbBatchNum, err := hermezDb.GetBatchNoByL2Block(entry.L2BlockNumber)
+					if err != nil {
+						return err
+					}
+
+					if entry.BatchNumber != dbBatchNum {
+						// if the bath number mismatches, it means that we need to trigger an unwinding of blocks
+						log.Warn(fmt.Sprintf("[%s] Batch number mismatch detected", logPrefix),
+							"block", entry.L2BlockNumber, "ds batch", entry.BatchNumber, "db batch", dbBatchNum)
+
+						ancestorBlockNum, ancestorBlockHash, err := findCommonAncestor(eriDb, hermezDb, dsQueryClient, entry.L2BlockNumber)
+						if err != nil {
+							return err
+						}
+
+						log.Info(fmt.Sprintf("[%s] found ancestor block %d (%s)", logPrefix, ancestorBlockNum, ancestorBlockHash))
+
+						unwindBlockNum, unwindBlockHash, batchNum, err := resolveUnwindBlock(eriDb, hermezDb, ancestorBlockNum, ancestorBlockHash)
+						if err != nil {
+							return err
+						}
+
+						log.Warn(fmt.Sprintf("[%s] Unwinding to block %d (%s)", logPrefix, unwindBlockNum, unwindBlockHash))
+						err = stages.SaveStageProgress(tx, stages.HighestSeenBatchNumber, batchNum-1)
+						if err != nil {
+							return err
+						}
+						u.UnwindTo(unwindBlockNum, unwindBlockHash)
+
+						cfg.dsClient.Stop()
+
+						return nil
```

Comment:
This is similar to the [code block below](https://github.com/0xPolygonHermez/cdk-erigon/pull/886/files#diff-477089124b274dd75c833b38e783d4cadf9ccaeb6c923e978c2cb092b69a846dR414-R435) where parent where block hash doesn't match. Can we refactor it into a function and use this function in both places?

Reply:
Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/6df39634229ad2adeb2f6157f0b7f46908e1f326